dali/kernels/signal/dct/dct_gpu.cu

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/signal/dct/dct_gpu.h"
+#include <cmath>
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/signal/dct/table.h"
+#include "dali/core/tensor_shape_print.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace dct {
+
+// The kernel processes data with the shape reduced to 3D.
+// Transform is applied over the middle axis.
+template <typename OutputType, typename InputType>
+__global__ void ApplyDct(const typename Dct1DGpu<OutputType, InputType>::SampleDesc *samples,
+                         const BlockDesc<3> *blocks)  {
+  int bid = blockIdx.x + gridDim.x * (blockIdx.y + gridDim.y * blockIdx.z);
+  auto block = blocks[bid];
+  const auto &sample = samples[block.sample_idx];
+  ivec3 in_stride = sample.in_stride;
+  ivec3 out_stride = sample.out_stride;
+
+  for (int z = block.start.z + threadIdx.z; z < block.end.z; z += blockDim.z) {
+    for (int y = block.start.y + threadIdx.y; y < block.end.y; y += blockDim.y) {
+      const OutputType *cos_row = sample.cos_table + sample.input_length * y;
+      for (int x = block.start.x + threadIdx.x; x < block.end.x; x += blockDim.x) {
+        int output_idx = dot(out_stride, ivec3{z, y, x});
+        const InputType *input = sample.input + dot(in_stride, ivec3{z, 0, x});
```

Comment:
This is wasting 2 integer multiplications per loop, since the innermost stride is invariably 1.